python/src/iceberg/avro/resolver.py

```diff
@@ -0,0 +1,150 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from functools import singledispatch
+from typing import (
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+)
+
+from iceberg.avro.reader import (
+    ConstructReader,
+    ListReader,
+    MapReader,
+    NoneReader,
+    OptionReader,
+    Reader,
+    StructReader,
+    primitive_reader,
+)
+from iceberg.schema import Schema, visit
+from iceberg.types import (
+    BinaryType,
+    IcebergType,
+    IntegerType,
+    ListType,
+    LongType,
+    MapType,
+    PrimitiveType,
+    StringType,
+    StructType,
+)
+
+
+class ResolveException(Exception):
+    pass
+
+
+@singledispatch
+def resolve(write_schema, read_schema) -> Reader:
+    """This resolves the write and read schema
+
+    The function traverses the schema in post-order fashion
+
+     Args:
+         write_schema (Schema | IcebergType): The write schema of the Avro file
+         read_schema (Schema | IcebergType): The requested read schema which is equal or a subset of the write schema
+
+     Raises:
+         NotImplementedError: If attempting to resolve an unrecognized object type
+    """
+    raise NotImplementedError("Cannot resolve non-type: %s" % write_schema)
+
+
+@resolve.register(Schema)
+def _(write_schema: Schema, read_schema: Schema) -> Reader:
+    """Visit a Schema and starts resolving it by converting it to a struct"""
+    return resolve(write_schema.as_struct(), read_schema.as_struct())
+
+
+@resolve.register(StructType)
+def _(write_struct: StructType, read_struct: StructType) -> Reader:
+    """Iterates over the write schema, and checks if the field is in the read schema"""
+    results: List[Tuple[Optional[int], Reader]] = []
+
+    read_fields = {field.name: (pos, field) for pos, field in enumerate(read_struct.fields)}
+
+    for write_field in write_struct.fields:
+        if write_field.name in read_fields:
+            read_pos, read_field = read_fields[write_field.name]
+            result_reader = resolve(write_field.field_type, read_field.field_type)
+        else:
+            read_pos = None
+            result_reader = visit(write_field.field_type, ConstructReader())
+        result_reader = result_reader if write_field.required else OptionReader(result_reader)
+        results.append((read_pos, result_reader))
+
+    write_fields = {field.name: field for field in write_struct.fields}
+    for pos, read_field in enumerate(read_struct.fields):
+        if read_field.name not in write_fields:
+            if read_field.required:
+                raise ResolveException(f"{read_field} is in not in the write schema, and is required")
```

Comment:
I recommend referring to the "file schema" rather than the "write" or "writer" schema. I've found that makes the most sense to people when reading error messages because they know where it's coming from "write schema" sounds like they're writing something, which can be confusing.

Reply:
I agree 👍🏻